diff --git a/codina/codina-vendor-blobs.mk b/codina/codina-vendor-blobs.mk
index 6dafb5b..ea7b0db 100644
--- a/codina/codina-vendor-blobs.mk
+++ b/codina/codina-vendor-blobs.mk
@@ -158,7 +158,6 @@ PRODUCT_COPY_FILES += \
     $(LOCAL_PATH)/system/lib/libsterc.so:system/lib/libsterc.so \
     $(LOCAL_PATH)/system/lib/libtee.so:system/lib/libtee.so \
     $(LOCAL_PATH)/system/lib/libtrace.so:system/lib/libtrace.so \
-    $(LOCAL_PATH)/system/lib/libstlport.so:system/lib/libstlport.so \
     $(LOCAL_PATH)/system/lib/libste_aiq_tools.so:system/lib/libste_aiq_tools.so
 
 # OMX
