diff --git a/include/hardware/audio.h b/include/hardware/audio.h
index c3cb740..6aef16a 100644
--- a/include/hardware/audio.h
+++ b/include/hardware/audio.h
@@ -669,6 +669,19 @@ struct audio_hw_device {
     void (*close_output_stream)(struct audio_hw_device *dev,
                                 struct audio_stream_out* stream_out);
 
+#ifdef STE_SAMSUNG_HARDWARE
+    /** This method creates and opens the audio hardware output
+     *  for broadcast stream */
+    int (*open_broadcast_stream)(struct audio_hw_device *dev, uint32_t devices,
+                                 int format, uint32_t channels,
+                                 uint32_t sample_rate,
+                                 uint32_t audio_source,
+                                 struct audio_broadcast_stream **out);
+
+    void (*close_broadcast_stream)(struct audio_hw_device *dev,
+                                   struct audio_broadcast_stream *out);
+#endif
+
     /** This method creates and opens the audio hardware input stream */
     int (*open_input_stream)(struct audio_hw_device *dev,
                              audio_io_handle_t handle,
